src/lib.rs

```diff
@@ -293,6 +225,119 @@ The Mobile Releases Team
+            if percentage >= 100.0 {
+                crash.insert(
+                    "percentage".to_string(),
+                    json!(format!("{:.2}%", percentage)),
+                );
+                crash.insert("threshold_exceeded".to_string(), json!(threshold));
+            } else {
+                // remove this to get rid of displaying crashes below threshold
+                crash.insert(
+                    "percentage".to_string(),
+                    json!(format!("{:.2}%", percentage)),
+                );
+            }
+            crash.insert("threshold".to_string(), json!(threshold));
+        }
+    }
+
+    fn add_operating_system_information(
+        &self,
+        crash_data: &mut serde_json::Map<String, serde_json::Value>,
+    ) {
+        let value = &mut crash_data["errorGroups"];
+        let all_crashes: &mut Vec<serde_json::Value> = value.as_array_mut().unwrap();
+
+        // object: Object<Map<String, Value>>
+        for object in all_crashes.iter_mut() {
+            let crash = object.as_object_mut().unwrap();
+            let oses = crash["operating_systems"].as_array().unwrap();
+            let crash_amount_devices_overall = crash["deviceCount"].as_f64().unwrap();
+            let mut formatted = oses
+                .iter()
+                .filter_map(|os| {
+                    let crash_amount_os = os["errorCount"].as_f64().unwrap();
+                    let percentage =
+                        (crash_amount_os / crash_amount_devices_overall) as f64 * 100.0;
+                    if percentage > 5.0 {
+                        let os_string = os["operatingSystemName"].as_str().unwrap();
+                        Some(format!(
+                            "{}: {} crashes ({:.2}%) ",
+                            os_string, crash_amount_os, percentage
+                        ))
+                    } else {
+                        None
+                    }
+                })
+                .fold(String::new(), |mut formatted, sub| {
+                    formatted += sub.as_str();
+                    formatted += "| ";
+                    formatted
+                });
+
+            // remove trailing "| ", as the previous 'fold' section can't know when
+            // to not put the suffix.
+            formatted.truncate(formatted.len() - 3);
+            let amount_of_affected_oses_shown = formatted.matches("|").count();
+            if amount_of_affected_oses_shown < crash["deviceCount"].as_u64().unwrap() as usize {
+                formatted += " and more"
+            }
+            crash.insert("operatingSystemName".to_string(), json!(formatted));
+        }
+    }
+    fn add_arithmetic_mean(&self, crash_data: &mut serde_json::Map<String, serde_json::Value>) {
+        // sum of all crashes / amount of crashes
+        // {"errorGroups": Array([…])}
+        let value = crash_data.get_mut("errorGroups").unwrap();
```

Comment:
This force `unwrap` is the problem right? It's like doing `Fooba!` in Swift, it will crash if `errorGroups` is nil correct? I think there are more graceful ways to handle this in Rust, but I haven't looked at Rust on the last two years, so I can't remember how. Perhaps @q231950 has some advice?

Reply:
Hm, in general that is true. In our case, I think, [this struct](https://github.com/xing/recrep/blob/master/src/model/crash_list.rs#L6) in combination with the JSON parsing takes care that the string `errorGroups` is present.